packages/opentelemetry-tracer-web/src/StackScopeManager.ts

```diff
@@ -0,0 +1,161 @@
+ */
+
+import { ScopeManager } from '@opentelemetry/scope-base';
+
+export class StackScopeManager implements ScopeManager {
+  private currentUid: number = 0;
+  private enabled = false;
+
+  public scopes: { [uid: number]: unknown } = Object.create(null);
+  public scopesStack: any[] = [];
+
+  constructor() {}
+
+  private activateScope(scope: any, uid: number) {
+    if (typeof scope === 'object') {
+      scope.uid = uid;
+    }
+    this.scopes[uid] = scope;
+    this.scopesStack.push(scope);
+  }
+
+  private bindFunction<T extends Function>(target: T, scope?: unknown): T {
+    const manager = this;
+    const contextWrapper: any = function(...args: unknown[]) {
+      return manager.with(scope, () => target.apply(scope, args));
+    };
+    Object.defineProperty(contextWrapper, 'length', {
+      enumerable: false,
+      configurable: true,
+      writable: false,
+      value: target.length,
+    });
+    return contextWrapper as any;
+  }
+
+  private createNewUid(): number {
+    return (this.currentUid = this.currentUid + 1);
+  }
+
+  private createRootScope(scope: any) {
+    const uid = typeof scope.uid === 'number' ? scope.uid : this.createNewUid();
+    this.activateScope(scope, uid);
+  }
+
+  private lastScope(): unknown {
+    return this.scopesStack[this.scopesStack.length - 1] || undefined;
+  }
+
+  private removeFromStack(uid: number) {
+    const index = this.scopesStack.lastIndexOf(this.scopes[uid]);
+    // don't remove root scope
+    if (index > 0) {
+      this.scopesStack.splice(index, 1);
+    }
+  }
+
+  private removeScope(uid: number) {
+    this.removeFromStack(uid);
+    const index = this.scopesStack.indexOf(this.scopes[uid]);
+
+    // don't remove root scope
+    if (index < 0 && this.scopes[uid] !== window) {
+      delete this.scopes[uid];
+    }
+  }
+
+  active(): unknown {
+    return this.lastScope();
+  }
+
+  bind<T>(target: T | any, scope?: unknown): T {
+    // if no specific scope to propagate is given, we use the current one
+    if (scope === undefined) {
+      scope = this.active();
+    }
+    if (typeof target === 'function') {
+      return this.bindFunction(target, scope);
+    }
+    return target;
+  }
+
+  disable(): this {
+    this.scopes = {};
+    this.scopesStack = [];
+    this.enabled = false;
+    return this;
+  }
+
+  enable(): this {
+    if (this.enabled) {
+      return this;
+    }
+    this.enabled = true;
+    this.createRootScope(window);
+    return this;
+  }
+
+  with<T extends (...args: unknown[]) => ReturnType<T>>(
+    scope: any,
+    fn: any
+  ): ReturnType<T> {
+    if (typeof scope === 'undefined' || scope === null) {
+      scope = window;
+    }
+    const uid = typeof scope.uid === 'number' ? scope.uid : this.createNewUid();
+    this.activateScope(scope, uid);
+
+    try {
+      return fn.apply(scope);
+    } catch (err) {
+      throw err;
+    } finally {
+      this.removeScope(uid);
+    }
+  }
+
+  // @TODO make async
```

Comment:
Per comment above, what would you think about making the async aspects of the scope manager be a totally different class? I think making Zone.js dependency optional for users will be a nice benefit since it's a lot of code and some people may just want tracing of the document load or simple XHRs, and so async context propagation wouldn't be needed for them.

Reply:
We could for sure allow user to pass a custom `ScopeManager` to the `tracer-web` and build a `ZoneScopeManager`

Reply:
I have removed comment and as you said creating a separate "ZoneScopeManager" is a way to go